### DISCUSSION TITLE
Available simulators/emulators?

### DISCUSSION THREAD

**Discussion by jforge** [digitalpetri/ethernet-ip#75] [General]
Is there any containerized ethernet-ip simulator/emulator available, to at least mimic real devices for faster and more reliable integration testing?

**Comment by kevinherron** [comment]
If you find one let me know 🙃

**Reply by jforge** [reply]
Done and - as always - easier that i thought. I didn't saw the relation between a repo called "cpppo" and EtherNet/IP, some AI assistant helped out somehow referring to it: https://github.com/pjkundert/cpppo

Skipping recompile and using their old docker images was enough for my purposes:

`docker run -p 44818:44818 --name cpppo-scada -d cpppo/scada`

Test with Node-RED and the node-contrib "cip-ethernet-ip":
`[{"id":"034da66f648b707b","type":"eth-ip in","z":"c7649e7a83815dda","endpoint":"f3910fcff11a13a1","mode":"all","variable":"","program":"","name":"","x":310,"y":1000,"wires":[["acc646da6ad9d46d"]]},{"id":"f3910fcff11a13a1","type":"eth-ip endpoint","address":"<some-host-ip>","slot":"0","cycletime":"500","name":"local-cpppo-server","vartable":{"":{"SCADA":{"type":"DINT"}}}}]
`

**Reply by kevinherron** [reply]
Nice. Some day I'd like to do a big 2.0 refactor of this library and explore adding adapter/server support.
